---
title: "D. Using Single-Cell Data into Bioconductor Work Flows"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{D. Using Single-Cell Data into Bioconductor Work Flows}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE)
```

This article explores more direct manipulation of single cell objects
and their integration into *Seurat*, *R* and *Bioconductor* work
flows. We illustrate the basics of object manipulation provided by
existing packages, and then hint at how one might start to work with
data 'outside the box' provided by these packages.  This can be
important for both routine (e.g., creating a subset of data based on
gene set membership) or advanced (e.g., implementing a new analysis
method) tasks. The focus is on _Bioconductor_'s `SingleCellExperiment`
object becuase of the rich supporting infastructure (e.g., annotation
resources or the powerful [GenomicRanges][] infrastructure) within
_Bioconductor_; similar operations can also be accomplished with the
Seurat object.

[GenomicRanges]: https://bioconductor.org/packages/GenomicRanges

We start by loading packages we'll use today.

```{r setup, message = FALSE}
library(HCABiocTraining)

## single cell data representation in R
library(Seurat)               # Seurat representation
library(SingleCellExperiment) # Bioconductor representation

## data access
library(cellxgenedp)

## general programming tools
library(dplyr)
library(ggplot2)
library(plotly)
```

We illustrate these steps with data from CELLxGENE retrieved in the
previous article. For our own data we would need to proceed from a
simple count matrix produced by Cell Ranger (etc.) through careful
quality control, normalization, data integration / batch coorrection,
etc., as illustrated in a different article.

# Exploring Objects

## Seurat

Based on [Getting Started with Seurat][] and the [Seurat - Guided
Clustering Tutorial][Seurat tutorial]. 

[Getting Started with Seurat]: https://satijalab.org/seurat
[Seurat tutorial]: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html

Retrieve the Seurat object

```{r}
## use a known dataset ID, discovered previously..
dataset <- "de985818-285f-4f59-9dbd-d74968fddba3"
training_cxg_dataset(dataset)

## download (or retieve from the local cache) the file
seurat_file <-
    files() |>
    dplyr::filter(filetype == "RDS", dataset_id == dataset) |>
    files_download(dry.run = FALSE)

seurat <- readRDS(seurat_file)
```

Retrieve the annotation on each cell using `[[]]`, e.g.,

```{r}
seurat[[]] |>
    as_tibble() |>
    dplyr::count(self_reported_ethnicity)

seurat[[]] |>
    as_tibble() |>
    dplyr::count(cell_type)
```

Recreate a simple UMAP

```{r}
UMAPPlot(seurat, group.by = "cell_type")
```

I remove the `seurat` object from the session, in hopes of freeing up
a bit of memory.

```{r}
rm(seurat)
```

## SingleCellExperiment

Retreive the `.h5ad` file

```{r}
## download (or retieve from the local cache) the file
dataset <- "de985818-285f-4f59-9dbd-d74968fddba3"
h5ad_file <-
    files() |>
    dplyr::filter(filetype == "H5AD", dataset_id == dataset) |>
    files_download(dry.run = FALSE)

## import into R as a SingleCellExperiment
h5ad <- training_read_h5ad_as_sce(h5ad_file)
```

## Basic Manipulation

The following is based on [Orchestrating Single-Cell Analysis with
Bioconductor][OSCA]. This resource is separated into introductory,
basic, advanced, and multi-sample sections, with a collection of
workflows illustrating use. A good place to start is with a workflow
to get a feel for data analysis, and then to refer back to earlier
sections for more detailed operations or understanding.

[OSCA]: http://bioconductor.org/books/release/OSCA.basic/

Displaying the `SingleCellExperiment`...

```{r}
h5ad
```

...suggests the data available and how to access it -- there are 
`r nrow(h5ad)` genes and `r ncol(h5ad)` cells. The 'raw' data include
a gene x cell count matricies (`assay()`) with annotations on the
genes (`rowData()`) and columns (`cellData()`), and with
reduced-dimension summaries.

For instance, the cell (column) annotations are easily
accessible and summarized as

```{r}
colData(h5ad) |>
    as_tibble()
```
```{r}
colData(h5ad) |>
    as_tibble() |>
    group_by(donor_id, self_reported_ethnicity) |>
    dplyr::count()
```
```{r}
colData(h5ad) |>
    as_tibble() |>
    dplyr::count(cell_type, sort = TRUE)
```

Many common operations have been standardized in a number of
_Bioconductor_ packages. For instance, use functionality from the
[scater][] package to visualize the UMAP present in the object,
coloring by the `cell_type` column data annotation.

```{r}
scater::plotReducedDim(h5ad, "X_umap", color_by = "cell_type")
```

[scater]: https://bioconductor.org/packages/scater

# Outside the Box

This section uses the *SingleCellExperiment*; similar operations can
be performed on *Seurat* objects.

It very easy to find out information about the samples included in the
study, e.g., the individual donors, their ethnicty, and family history
of breast cancer (the specific information available depends on the
data submitted by the original investigator).

```{r}
colData(h5ad) |>
    as_tibble() |>
    dplyr::count(
        donor_id, self_reported_ethnicity, family_history_breast_cancer
    )
```


The actual feature x sample counts are available and easily
manipulated, as illustrated by this plot of the distribution of reads
per cell...

```{r}
reads_per_cell <-
    h5ad |>
    ## retrieve the matrix of gene x cell counts
    assay("counts", withDimnames = FALSE) |>
    ## calculate the column sums, i.e., reads mapped to each cell
    colSums()

hist(log10(reads_per_cell))
```

...or to remove genes with non-zero counts

```{r}
reads_per_gene <-
    h5ad |>
    assay("counts", withDimnames = FALSE) |>
    rowSums()

table(reads_per_gene != 0)

h5ad[reads_per_gene != 0,]
```

## Visualization

The previous section created a static visualization from the 'UMAP'
reduced dimension representation using `scater::plotReducedDim()`:

It can be helpful to do this 'by hand' to illustrate how one can work
directly with *SingleCellExperiment* objects. First load [ggplot2][]

[ggplot2]: https://cran.r-project.org/package=ggplot2

```{r}
library(ggplot2)
```

Then create a tibble containing information about the UMAP, as well as
cell (column) annotations

```{r}
umap <-
    as_tibble(reducedDim(h5ad, "X_umap")) |>
    bind_cols(
        cell_type = h5ad$cell_type,
        donor_id = h5ad$donor_id,
        self_reported_ethnicity = h5ad$self_reported_ethnicity,
        colname = colnames(h5ad) # unique identifier
    )
```

Visualize the two-dimensional UMAP summary, coloring by cell type

```{r}
ggplot(umap) +
    aes(x = V1, y = V2, color = cell_type) +
    geom_point(pch = ".")
```

The [plotly][] package provides a very convenient way to make this an
interactive plot

[plotly]: https://cran.r-project.org/package=plotly

```{r}
library(plotly)

plot_ly(
    umap,
    x = ~ V1, y =  ~ V2, color = ~ cell_type,
    type = "scatter", mode = "markers", opacity = .8,
    marker = list(symbol = "circle-open", line = list(width = 1))
)
```

A suprisingly straight-forward (but admittedly advanced) helper
function uses [shiny][] to allows us to write an application to select
data we are interested in... check it out!

[shiny]: https://cran.r-project.org/package=shiny

```{r, eval = FALSE}
result <- training_cell_viewer(umap) # select some cells of interest
result
## subset the SingleCellExperiment to just those cells
h5ad[, result$colname]
```

## Gene Sets

The single cell assay detected expression of `r nrow(h5ad)` genes, but
we may often be interested in only a set of these. Such gene sets
might be defined by our own research interests, or may be defined as
'community standards'.

Check out [MSigDB][] -- a collection of gene sets! Focus on the
Halmarks of Cancer gene sets 'H: hallmark gene sets'.  **NOTE** MSigDB
requires registration, which is triggered by downloading a gene set
manually; do that first, even though we use automatically downloaded
gene sets.

The helper function `training_hallmarks()` downloads the 'Hallmarks of
Cancer' gene sets from MSigDb, translating the 'Entrez' gene
identifiers in the gene sets to the 'Ensembl' identifiers
(`rownames(h5ad)`) in our data.

```{r}
hallmarks <- training_hallmarks()
hallmarks

## gene sets and their counts
hallmarks |> dplyr::count(set)
```

Suppose we are interested in the `HALLMARK_P53_PATHWAY` gene set

```{r}
p53_gene_set <-
    hallmarks |>
    filter(set == "HALLMARK_P53_PATHWAY")
nrow(p53_gene_set)
```

We can create a subset of our data with just these genes (some
identifiers in `p53_gene_set` are not present in our observed data...)

```{r}
p53_rows <- rownames(h5ad) %in% pull(p53_gene_set, "gene")
h5ad_p53 <- h5ad[p53_rows,]
h5ad_p53
```

These steps illustrate some of the straight-forward ways of
integrating scRNASeq analysis into larger bioinformatic work flows.

[MSigDB]: https://www.gsea-msigdb.org/gsea/msigdb/
