<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="HCABiocTraining">
<title>C. Single-Cell Sequence Analysis • HCABiocTraining</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="C. Single-Cell Sequence Analysis">
<meta property="og:description" content="HCABiocTraining">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">HCABiocTraining</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.14</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/a-intro-to-r.html">A. Introduction to R</a>
    <a class="dropdown-item" href="../articles/b-hca-data-access.html">B. Accessing HCA Data from R / Bioconductor</a>
    <a class="dropdown-item" href="../articles/c-sc-analysis.html">C. Single-Cell Sequence Analysis</a>
    <a class="dropdown-item" href="../articles/d-objects.html">D. Using Single-Cell Data into Bioconductor Work Flows</a>
    <a class="dropdown-item" href="../articles/e-directions.html">E. Spatial, Multi-Modal and Extended Analysis</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mtmorgan/HCABiocTraining" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>C. Single-Cell Sequence Analysis</h1>
            
      
      
      <div class="d-none name"><code>c-sc-analysis.Rmd</code></div>
    </div>

    
    
<p>In the previous article we saw how to make HCA / CELLxGENE data
available as <code>SingleCellExperiment</code> or <code>Seuart</code>
objects. From here there are rich, well-documented workflows for
subsequent analysis. We’ll point to these resources and outline overall
workflows, but will only perform illustrative calculations.</p>
<p>Some steps require both computing time and resources that aren’t
suitable for workshop presentation so we focus on a ‘classic’ data set,
the 10x Genomics Peripheral Blood Mononuclear Cells (PBMC) dataset
conssists of 2,700 cells. Analysis of data sets with say 50,000 cells
might require 32- or 64 GB of memory and take up to an hour for an
end-to-end analysis; one way to access a computer with the much memory
might be through a commercial (Google, Azure, AWS) cloud, or on a
research network like the <a href="https://anvilproject.org" class="external-link">AnVIL</a>
computational cloud.</p>
<div class="section level2">
<h2 id="seurat">Seurat<a class="anchor" aria-label="anchor" href="#seurat"></a>
</h2>
<p>Based on <a href="https://satijalab.org/seurat" class="external-link">Getting Started with
Seurat</a> and the <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">Seurat
- Guided Clustering Tutorial</a>. The guided tutorial covers</p>
<ul>
<li>Data input</li>
<li>Quality Control</li>
<li>Normalization</li>
<li>Identification of highly variable features</li>
<li>Scaling the data</li>
<li>Linear dimensional reduction</li>
<li>Clusters</li>
<li>Non-linear dimensional reduction (e.g., UMAP)</li>
<li>Differential expression</li>
<li>Cell type assignment</li>
</ul>
<p>To illustrating working with <code>Seurat</code>, we take the first
few steps through this workflow.</p>
<div class="section level3">
<h3 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h3>
<p>Here we re-capitulate some of the initial steps in the <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">Seurat
tutorial</a>.</p>
<p>Data retrieval. Here we retrieve the raw data to a local cache. It is
distributed as a compressed ‘tar’ archive, so we <code><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar()</a></code> it
to a temporary location.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## download to local cache</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://cf.10xgenomics.com/samples/cell/pbmc3k/"</span>,</span>
<span>    <span class="st">"pbmc3k_filtered_gene_bc_matrices.tar.gz"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pbmc3k_tar_gz</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcrpath</a></span><span class="op">(</span>rname <span class="op">=</span> <span class="va">url</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## 'untar' to a temporary location</span></span>
<span><span class="va">pbmc3k_directory</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span><span class="va">pbmc3k_tar_gz</span>, exdir <span class="op">=</span> <span class="va">pbmc3k_directory</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## location of the relevant data</span></span>
<span><span class="va">pbmc3k_data_dir</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pbmc3k_directory</span>, <span class="st">"filtered_gene_bc_matrices"</span>, <span class="st">"hg19"</span><span class="op">)</span></span></code></pre></div>
<p>Following the <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">Seurat
tutorial</a>, we read the data into a sparse matrix, and then into a
Seurat object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">## Attaching SeuratObject</span></span>
<span></span>
<span><span class="co">## read the 10x data as a sparse matrix</span></span>
<span><span class="va">pbmc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="va">pbmc3k_data_dir</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">pbmc.data</span><span class="op">)</span></span>
<span><span class="co">## [1] "dgCMatrix"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "Matrix"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pbmc.data</span><span class="op">)</span> <span class="co"># 32738 features (genes) x 2700 samples (cells)</span></span>
<span><span class="co">## [1] 32738  2700</span></span>
<span></span>
<span><span class="co">## create a Seurat object from the count matrix</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>    counts <span class="op">=</span> <span class="va">pbmc.data</span>,</span>
<span>    project <span class="op">=</span> <span class="st">"pbmc3k"</span>,</span>
<span>    <span class="co">## filter some features (present in less than 3 cells) and cells</span></span>
<span>    <span class="co">## (fewer than 200 features)</span></span>
<span>    min.cells <span class="op">=</span> <span class="fl">3</span>, min.features <span class="op">=</span> <span class="fl">200</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Warning: Feature names cannot have underscores ('_'), replacing with dashes</span></span>
<span><span class="co">## ('-')</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="co"># 13714 features x 2700 samples</span></span>
<span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 13714 features across 2700 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (13714 features, 0 variable features)</span></span>
<span></span>
<span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## Assay data with 13714 features for 2700 cells</span></span>
<span><span class="co">## First 10 features:</span></span>
<span><span class="co">##  AL627309.1, AP006222.2, RP11-206L10.2, RP11-206L10.9, LINC00115, NOC2L,</span></span>
<span><span class="co">## KLHL17, PLEKHN1, RP11-54O7.17, HES4</span></span></code></pre></div>
<p>The first step in the workflow is to perform quality control. Start
by identifying the mitochondrial genes. In this particular example,
mitchondrial genes are annotated based on the feature name.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## The [[ operator can add columns to object metadata. This is a great</span></span>
<span><span class="co">## place to stash QC stats</span></span>
<span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span></code></pre></div>
<p>Visualize some metrics…</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">pbmc</span>,</span>
<span>    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span><span class="op">)</span>,</span>
<span>    ncol <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="c-sc-analysis_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>…and use <em>ad hoc</em> criteria to filter features;
<code><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset()</a></code> is a base <em>R</em> function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>    <span class="va">pbmc</span>,</span>
<span>    subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">200</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2500</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc</span></span>
<span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 13714 features across 2638 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (13714 features, 0 variable features)</span></span></code></pre></div>
<p>The workflow continues from here…</p>
</div>
</div>
<div class="section level2">
<h2 id="bioconductor">Bioconductor<a class="anchor" aria-label="anchor" href="#bioconductor"></a>
</h2>
<p>The following is based on <a href="http://bioconductor.org/books/release/OSCA.basic/" class="external-link">Orchestrating
Single-Cell Analysis with Bioconductor</a>. This resource is separated
into introductory, basic, advanced, and multi-sample sections, with a
collection of workflows illustrating use. A good place to start is with
a workflow to get a feel for data analysis, and then to refer back to
earlier sections for more detailed operations or understanding.</p>
<p>We’ll follow the <a href="https://bioconductor.org/books/release/OSCA.workflows/unfiltered-human-pbmcs-10x-genomics.html" class="external-link">Unfiltered
human PBMCs</a> workflow for another PBMC data, but use the same dataset
as in the Seurat section earlier in this article.</p>
<div class="section level3">
<h3 id="workflow-1">Workflow<a class="anchor" aria-label="anchor" href="#workflow-1"></a>
</h3>
<p>Data retrieval is a outlined above. After <code><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar()</a></code>ing the
data <code>pbmc3k_data_dir</code> is the location to use for data
input.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DropletUtils</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html" class="external-link">read10xCounts</a></span><span class="op">(</span><span class="va">pbmc3k_data_dir</span>, col.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## as(&lt;dgTMatrix&gt;, "dgCMatrix") is deprecated since Matrix 1.5-0; do as(., "CsparseMatrix") instead</span></span>
<span></span>
<span><span class="va">pbmc3k</span></span>
<span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 32738 2700 </span></span>
<span><span class="co">## metadata(1): Samples</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(32738): ENSG00000243485 ENSG00000237613 ... ENSG00000215616</span></span>
<span><span class="co">##   ENSG00000215611</span></span>
<span><span class="co">## rowData names(2): ID Symbol</span></span>
<span><span class="co">## colnames(2700): AAACATACAACCAC-1 AAACATTGAGCTAC-1 ... TTTGCATGAGAGGC-1</span></span>
<span><span class="co">##   TTTGCATGCCTCAC-1</span></span>
<span><span class="co">## colData names(2): Sample Barcode</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre></div>
<p>Skipping to the quality control step, let’s identify mitochondrial
genes. In this data set, we can identify the mitochondrial genes by
looking at the <code>Symbol</code> column of <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/startsWith.html" class="external-link">startsWith</a></span><span class="op">(</span><span class="va">Symbol</span>, <span class="st">"MT-"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 13 × 2</span></span></span>
<span><span class="co">##    ID              Symbol </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ENSG00000198888 MT-ND1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> ENSG00000198763 MT-ND2 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ENSG00000198804 MT-CO1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ENSG00000198712 MT-CO2 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ENSG00000228253 MT-ATP8</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> ENSG00000198899 MT-ATP6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ENSG00000198938 MT-CO3 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ENSG00000198840 MT-ND3 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ENSG00000212907 MT-ND4L</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> ENSG00000198886 MT-ND4 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">11</span> ENSG00000198786 MT-ND5 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">12</span> ENSG00000198695 MT-ND6 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">13</span> ENSG00000198727 MT-CYB</span></span>
<span></span>
<span><span class="va">is_mitochondrial</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/startsWith.html" class="external-link">startsWith</a></span><span class="op">(</span><span class="st">"MT-"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Calculate per-cell quality control metrics…</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">pbmc3k</span>, subsets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Mito<span class="op">=</span><span class="va">is_mitochondrial</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">stats</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 2,700 × 6</span></span></span>
<span><span class="co">##      sum detected subsets_Mito_sum subsets_Mito_detected subsets_Mito_pe…¹ total</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>  <span style="text-decoration: underline;">2</span>421      781               73                    10             3.02   <span style="text-decoration: underline;">2</span>421</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>  <span style="text-decoration: underline;">4</span>903     <span style="text-decoration: underline;">1</span>352              186                    10             3.79   <span style="text-decoration: underline;">4</span>903</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>  <span style="text-decoration: underline;">3</span>149     <span style="text-decoration: underline;">1</span>131               28                     8             0.889  <span style="text-decoration: underline;">3</span>149</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>  <span style="text-decoration: underline;">2</span>639      960               46                    10             1.74   <span style="text-decoration: underline;">2</span>639</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>   981      522               12                     5             1.22    981</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>  <span style="text-decoration: underline;">2</span>164      782               36                     7             1.66   <span style="text-decoration: underline;">2</span>164</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>  <span style="text-decoration: underline;">2</span>176      783               83                    10             3.81   <span style="text-decoration: underline;">2</span>176</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>  <span style="text-decoration: underline;">2</span>260      790               70                     9             3.10   <span style="text-decoration: underline;">2</span>260</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>  <span style="text-decoration: underline;">1</span>276      533               15                     6             1.18   <span style="text-decoration: underline;">1</span>276</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>  <span style="text-decoration: underline;">1</span>103      550               32                     7             2.90   <span style="text-decoration: underline;">1</span>103</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 2,690 more rows, and abbreviated variable name ¹​subsets_Mito_percent</span></span></span>
<span></span>
<span><span class="va">high_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">stats</span><span class="op">$</span><span class="va">subsets_Mito_percent</span>, type<span class="op">=</span><span class="st">"higher"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">high_mito</span><span class="op">)</span></span>
<span><span class="co">## high_mito</span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##  2607    93</span></span></code></pre></div>
<p>…and use <code>high_mito</code> to create a subset of the results</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc3k</span><span class="op">[</span>, <span class="op">!</span><span class="va">high_mito</span><span class="op">]</span></span>
<span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 32738 2607 </span></span>
<span><span class="co">## metadata(1): Samples</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(32738): ENSG00000243485 ENSG00000237613 ... ENSG00000215616</span></span>
<span><span class="co">##   ENSG00000215611</span></span>
<span><span class="co">## rowData names(2): ID Symbol</span></span>
<span><span class="co">## colnames(2607): AAACATACAACCAC-1 AAACATTGAGCTAC-1 ... TTTGCATGAGAGGC-1</span></span>
<span><span class="co">##   TTTGCATGCCTCAC-1</span></span>
<span><span class="co">## colData names(2): Sample Barcode</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre></div>
<p>The workflow continues…!</p>
</div>
</div>
<div class="section level2">
<h2 id="brief-commentary">Brief Commentary<a class="anchor" aria-label="anchor" href="#brief-commentary"></a>
</h2>
<p>It is worthwhile to compare the <em>Seurat</em> and
<em>SingleCellExperiment</em> work flows.</p>
<p>Actually, a first question is ‘why not just have a single function
that takes the analysis from begining to end?’ Answers might be</p>
<ol style="list-style-type: decimal">
<li><p>It is educational to understand the data transformations and
their assummptions required in an analysis.</p></li>
<li><p>No two data sets are the same, so at any step some manipulation
(e.g., to identify mitochondrial genes), choice of algorithm, or
interpretation of results might be required.</p></li>
</ol>
<p>What about the difference between the <em>Seurat</em> and
<em>Bioconductor</em> workflows? In broad detail the analyses accomplish
the same ends. <em>Seurat</em> is more standardized and widely used,
while <em>Bioconductor</em> may beless standardized and more
flexible.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Morgan.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
